File: src/main/java/com/flexpoker/game/command/handlers/CreateGameCommandHandler.java
package com.flexpoker.game.command.handlers;

import javax.inject.Inject;

import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import com.flexpoker.framework.command.CommandHandler;
import com.flexpoker.framework.event.EventPublisher;
import com.flexpoker.game.command.commands.CreateGameCommand;
import com.flexpoker.game.command.factory.GameFactory;
import com.flexpoker.game.command.framework.GameEvent;
import com.flexpoker.game.command.repository.GameEventRepository;

@Component
public class CreateGameCommandHandler implements CommandHandler<CreateGameCommand> {

    private final GameFactory gameFactory;

    private final EventPublisher<GameEvent> eventPublisher;

    private final GameEventRepository gameEventRepository;

    @Inject
    public CreateGameCommandHandler(GameFactory gameFactory,
            EventPublisher<GameEvent> eventPublisher,
            GameEventRepository gameEventRepository) {
        this.gameFactory = gameFactory;
        this.eventPublisher = eventPublisher;
        this.gameEventRepository = gameEventRepository;
    }

    @Async
    @Override
    public void handle(CreateGameCommand command) {
        var game = gameFactory.createNew(command);
        game.fetchNewEvents().forEach(x -> gameEventRepository.save(x));
        game.fetchNewEvents().forEach(x -> eventPublisher.publish(x));
    }

}

File: src/main/kotlin/com/flexpoker/game/command/handlers/CreateGameCommandHandler.kt
package com.flexpoker.game.command.handlers

import com.flexpoker.framework.command.CommandHandler
import com.flexpoker.framework.event.EventPublisher
import com.flexpoker.game.command.aggregate.eventproducers.createGame
import com.flexpoker.game.command.commands.CreateGameCommand
import com.flexpoker.game.command.events.GameEvent
import com.flexpoker.game.command.repository.GameEventRepository
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import javax.inject.Inject

@Component
class CreateGameCommandHandler @Inject constructor(
    private val eventPublisher: EventPublisher<GameEvent>,
    private val gameEventRepository: GameEventRepository
) : CommandHandler<CreateGameCommand> {

    @Async
    override fun handle(command: CreateGameCommand) {
        val newEvents = createGame(command.gameName, command.numberOfPlayers, command.numberOfPlayersPerTable,
            command.createdByPlayerId, command.numberOfMinutesBetweenBlindLevels, command.numberOfSecondsForActionOnTimer)
        val eventsWithVersions = gameEventRepository.setEventVersionsAndSave(0, newEvents)
        eventsWithVersions.forEach { eventPublisher.publish(it) }
    }

}